### PR TITLE
1575 - implementation of ingest relationships and GET Relationships API

### PR DESCRIPTION
**Pull request by Vladysl**

### CODE REVIEW

File: odd-platform-specification/components.yaml [hunk 1]
```diff
@@ -3934,6 +3934,121 @@ components:
...
+          type: array
+          items:
+            $ref: '#/components/schemas/ERDRelationshipDetails'
+      required:
+        - items
+
+    ERDRelationshipDetails:
+      type: object
+      properties:
+        erd_relationships_id:
```
Comment (AndreyNenashev) [inline odd-platform-specification/components.yaml:4000]:
```suggestion
        erd_relationship_id:
```

File: odd-platform-specification/components.yaml [hunk 2]
```diff
@@ -3934,6 +3934,121 @@ components:
...
+          type: array
+          items:
+            $ref: '#/components/schemas/GraphRelationshipDetails'
+      required:
+        - items
+
+    GraphRelationshipDetails:
+      type: object
+      properties:
+        graph_relationships_id:
```
Comment (AndreyNenashev) [inline odd-platform-specification/components.yaml:4031]:
```suggestion
        graph_relationship_id:
```

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/mapper/GraphRelationshipMapper.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    private List<GraphRelationshipDetails> mapPojoListToDetailsList(final List<GraphRelationshipPojo> graphList) {
+        return graphList.stream()
+            .map(this::mapPojoToDetails)
+            .toList();
+    }
+
+    public GraphRelationshipDetails mapPojoToDetails(final GraphRelationshipPojo pojo) {
+        return new GraphRelationshipDetails()
+            .graphRelationshipsId(pojo.getId())
+            .isDirected(pojo.getIsDerected())
```
Comment (AndreyNenashev) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/mapper/GraphRelationshipMapper.java:38]:
```suggestion
            .isDirected(pojo.getIsDirected())
```

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    cardinality                varchar(256),
+
+    CONSTRAINT erd_relationship_fk_relationship FOREIGN KEY (relationship_id) REFERENCES relationship (id)
+);
+
+CREATE TABLE IF NOT EXISTS graph_relationship
+(
+    id                      bigserial PRIMARY KEY,
+    relationship_id         bigint,
+    is_derected             boolean,
```
Comment (AndreyNenashev) [inline odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql:30]:
```suggestion
    is_directed             boolean,
```

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql [hunk 5]
```diff
@@ -0,0 +1,36 @@
+CREATE TABLE IF NOT EXISTS relationship
```
Comment (AndreyNenashev) [inline odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql:1]:
```suggestion
CREATE TABLE IF NOT EXISTS relationships
```

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    relationship_oddrn         varchar(512) UNIQUE,
+    source_dataset_oddrn       varchar(512)       NOT NULL,
+    target_dataset_oddrn       varchar(512)       NOT NULL,
+    is_manualy_created         boolean,
+    relationship_status        SMALLINT,
+    last_ingested_at           TIMESTAMP WITHOUT TIME ZONE,
+    deleted_at                 TIMESTAMP WITHOUT TIME ZONE
+);
+
+CREATE TABLE IF NOT EXISTS erd_relationship
```
Comment (AndreyNenashev) [inline odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql:14]: I don't mind but it feels like `erd_relationship_details` is more semantically correct. Current naming pattern feels like `erd_relatatioships` table **inherits** `relationships` table which is not true

**Review by AndreyNenashev — CHANGES_REQUESTED** [review]

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    relationship_oddrn         varchar(512) UNIQUE,
+    source_dataset_oddrn       varchar(512)       NOT NULL,
+    target_dataset_oddrn       varchar(512)       NOT NULL,
+    is_manualy_created         boolean,
+    relationship_status        SMALLINT,
+    last_ingested_at           TIMESTAMP WITHOUT TIME ZONE,
+    deleted_at                 TIMESTAMP WITHOUT TIME ZONE,
+
+    CONSTRAINT relationship_fk_data_source FOREIGN KEY (data_source_id) REFERENCES data_source (id)
+);
```
Comment (AndreyNenashev) [inline odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql:15]: it's polymorphic one-to-one relation between `relationships ` table and `erd_relationship`, `graph_relationship` tables. We should add some `column_type` that defines it.

File: odd-platform-specification/components.yaml (hunk 1 above)
Comment (Vladysl) [reply odd-platform-specification/components.yaml:3992]: done

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/mapper/GraphRelationshipMapper.java (hunk 3 above)
Comment (Vladysl) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/mapper/GraphRelationshipMapper.java]: done

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql (hunk 5 above)
Comment (Vladysl) [reply odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql]: done

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql (hunk 6 above)
Comment (Vladysl) [reply odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql]: done

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql (hunk 4 above)
Comment (Vladysl) [reply odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql]: done

File: odd-platform-specification/components.yaml (hunk 2 above)
Comment (Vladysl) [reply odd-platform-specification/components.yaml]: done

File: odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql (hunk 7 above)
Comment (Vladysl) [reply odd-platform-api/src/main/resources/db/migration/V0_0_87__create_relation_tables.sql]: done

File: odd-platform-specification/components.yaml (hunk 1 above)
Comment (Vladysl) [reply odd-platform-specification/components.yaml]: done

**Review by AndreyNenashev — CHANGES_REQUESTED** [review]

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveRelationshipsRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+        final SelectConditionStep<Record> query = relationshipGenericQuery.where(RELATIONSHIPS.ID.eq(relationshipId));
+
+        return jooqReactiveOperations.mono(query)
+            .map(r -> new RelationshipDto(r.into(RELATIONSHIPS).into(RelationshipsPojo.class),
+                r.into(srcDataEntity).into(DataEntityPojo.class),
+                r.into(trgtDataEntity).into(DataEntityPojo.class),
+                r.into(ERD_RELATIONSHIP_DETAILS).into(ErdRelationshipDetailsPojo.class),
+                r.into(GRAPH_RELATIONSHIP).into(GraphRelationshipPojo.class))
+            );
```
[the remark this answers is not in the document]
Comment (Vladysl) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveRelationshipsRepositoryImpl.java]: updated

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveRelationshipsRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+            .on(GRAPH_RELATIONSHIP.RELATIONSHIP_ID.eq(relationshipCTE.field(RELATIONSHIPS.ID)));
+
+        final ResultQuery<Record> resultQuery = switch (type) {
+            case ERD -> generalQuery
+                .where(relationshipCTE.field(RELATIONSHIPS.RELATIONSHIP_TYPE).eq(RelationshipTypeDto.ERD.name()))
+                .groupBy(groupByFields);
+            case GRAPH -> generalQuery
+                .where(relationshipCTE.field(RELATIONSHIPS.RELATIONSHIP_TYPE).eq(RelationshipTypeDto.GRAPH.name()))
+                .groupBy(groupByFields);
+            default -> generalQuery.groupBy(groupByFields);
```
[the remark this answers is not in the document]
Comment (Vladysl) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveRelationshipsRepositoryImpl.java]: updated

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveRelationshipsRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+            .collectList()
+            .flatMap(record -> jooqQueryHelper.pageifyResult(
+                record,
+                r -> new RelationshipDto(
+                    jooqRecordHelper.remapCte(r, relationshipCTE.getName(), RELATIONSHIPS)
+                        .into(RelationshipsPojo.class),
+                    r.into(srcDataEntity).into(DataEntityPojo.class),
+                    r.into(trgtDataEntity).into(DataEntityPojo.class),
+                    r.into(ERD_RELATIONSHIP_DETAILS).into(ErdRelationshipDetailsPojo.class),
+                    r.into(GRAPH_RELATIONSHIP).into(GraphRelationshipPojo.class)),
```
[the remark this answers is not in the document]
Comment (Vladysl) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveRelationshipsRepositoryImpl.java]: updated

**Review by AndreyNenashev — CHANGES_REQUESTED** [review]

File: odd-platform-specification/openapi.yaml [hunk 11]
```diff
@@ -3811,4 +3829,41 @@ paths:
...
+        '200':
+          description: OK
+          content:
+            application/json:
+              schema:
+                $ref: './components.yaml/#/components/schemas/DatasetRelationshipList'
+      tags:
+        - relationship
+
+  /api/relationships/{relationships_id}:
```
[the remark this answers is not in the document]
Comment (Vladysl) [reply odd-platform-specification/openapi.yaml]: done

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/mapper/RelationshipMapper.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+import org.opendatadiscovery.oddplatform.dto.RelationshipDto;
+import org.opendatadiscovery.oddplatform.dto.RelationshipTypeDto;
+import org.opendatadiscovery.oddplatform.utils.Page;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class RelationshipMapper {
+    private final ErdRelationshipMapper erdRelationshipMapper;
+    private final GraphRelationshipMapper graphRelationshipMapper;
```
Comment (AndreyNenashev) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/mapper/RelationshipMapper.java:22]: I think we don't need that mapper to be dependant of ERD or Graph cases. For relationships page we need just `Relationship` object without details.  For details page FE will sequently request relationship and after that ERD or Graph details. So I think we should decompose this mapper

**Review by AndreyNenashev — CHANGES_REQUESTED** [review]

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveERDRelationshipsRepositoryImpl.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    public ReactiveERDRelationshipsRepositoryImpl(final JooqReactiveOperations jooqReactiveOperations,
+                                                  final JooqQueryHelper jooqQueryHelper,
+                                                  final JooqRecordHelper jooqRecordHelper) {
+        super(jooqReactiveOperations, jooqQueryHelper, ERD_RELATIONSHIP_DETAILS, ErdRelationshipDetailsPojo.class);
+
+        this.jooqRecordHelper = jooqRecordHelper;
+    }
+
+    @Override
+    public Mono<List<ErdRelationshipDetailsPojo>> findERDSByRelationIds(final List<Long> relationshipId) {
```
Comment (AndreyNenashev) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/reactive/ReactiveERDRelationshipsRepositoryImpl.java:34]:
```suggestion
    public Mono<List<ErdRelationshipDetailsPojo>> findERDSByRelationIds(final List<Long> relationshipIds) {
```

**Review by AndreyNenashev — DISMISSED** [review]

**Review by AndreyNenashev — APPROVED** [review]

**Pull request merged**